docs/architecture/adr-038-state-listening.md

```diff
@@ -182,349 +137,609 @@ func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
-    // HaltAppOnDeliveryError whether or not to halt the application when delivery of massages fails
-    // in ListenBeginBlock, ListenEndBlock, ListenDeliverTx. When `false, the app will operate in fire-and-forget mode.
-    // When `true`, the app will gracefully halt and stop the running node. Uncommitted blocks will
-    // be replayed to all listeners when the node restarts and all successful listeners that received data
-    // prior to the halt will receive duplicate data. Whether or not a listener operates in a fire-and-forget mode
-    // is determined by the listener's configuration property `halt_app_on_delivery_error = true|false`.
-    HaltAppOnDeliveryError() bool
-}
-
-// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
-type StreamingService interface {
-    // Stream is the streaming service loop, awaits kv pairs and writes them to a destination stream or file
-    Stream(wg *sync.WaitGroup) error
-    // Listeners returns the streaming service's listeners for the BaseApp to register
-    Listeners() map[types.StoreKey][]store.WriteListener
-    // ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
-    ABCIListener
-    // Closer interface
-    io.Closer
+    // ListenCommit updates the steaming service with the latest Commit messages and state changes
+    ListenCommit(ctx types.Context, res abci.ResponseCommit, changeSet []store.StoreKVPair) error
 }
 ```
 
-#### BaseApp registration
+#### ABCI Event Hooks
 
-We will add a new method to the `BaseApp` to enable the registration of `StreamingService`s:
+We will modify the `BeginBlock`, `EndBlock`, `DeliverTx` and `Commit` methods to pass ABCI requests and responses
+to any streaming service hooks registered with the `BaseApp`.
 
 ```go
-// SetStreamingService is used to set a streaming service into the BaseApp hooks and load the listeners into the multistore
-func (app *BaseApp) SetStreamingService(s StreamingService) {
-	// add the listeners for each StoreKey
-	for key, lis := range s.Listeners() {
-		app.cms.AddListeners(key, lis)
-	}
-	// register the StreamingService within the BaseApp
-	// BaseApp will pass BeginBlock, DeliverTx, and EndBlock requests and responses to the streaming services to update their ABCI context
-	app.abciListeners = append(app.abciListeners, s)
+func (app *BaseApp) BeginBlock(req abci.RequestBeginBlock) (res abci.ResponseBeginBlock) {
+
+    ...
+
+    // call the streaming service hook with the BeginBlock messages
+    if app.abciListener != nil {
+        ctx := app.deliverState.ctx
+        blockHeight := ctx.BlockHeight()
+        if app.stopNodeOnStreamingErr {
+            if err := app.abciListener.ListenBeginBlock(ctx, req, res); err != nil {
+                app.logger.Error("BeginBlock listening hook failed", "height", blockHeight, "err", err)
+                os.Exit(1)
```

Comment:
ref: https://github.com/cosmos/cosmos-sdk/pull/13473#discussion_r994080070

> Strictly speaking, for programs like Cosmos nodes, nothing should ever call panic or os.Exit or app.halt except func main. Practically speaking, you can't rely on panic to actually terminate the process: it's frequently used as an ersatz error reporting mechanism, which means there are lots of calls to recover that you can't reliably know about, which means your panic is very likely to be intercepted and squelched. If you want the process to terminate you better call os.Exit.

Reply:
Ok, that's fine. Although this is in non-tx-flow, so it will halt the node (we do this elsewhere and I'd rather be consistent, but I dont wanna die on that hill).